woocommerce-gateway-stripe.php

```diff
@@ -267,26 +275,51 @@ public function plugin_row_meta( $links, $file ) {
 			 * @version 4.0.0
 			 */
 			public function add_gateways( $methods ) {
-				if ( class_exists( 'WC_Subscriptions_Order' ) && function_exists( 'wcs_create_renewal_order' ) ) {
-					$methods[] = 'WC_Stripe_Subs_Compat';
-					$methods[] = 'WC_Stripe_Sepa_Subs_Compat';
+				if ( ! WC_Stripe_Features::is_upe_enabled() ) {
+					$methods[] = ( class_exists( 'WC_Subscriptions_Order' ) && function_exists( 'wcs_create_renewal_order' ) ) ? WC_Stripe_Subs_Compat::class : WC_Gateway_Stripe::class;
+					$methods[] = ( class_exists( 'WC_Subscriptions_Order' ) && function_exists( 'wcs_create_renewal_order' ) ) ? WC_Stripe_Sepa_Subs_Compat::class : WC_Gateway_Stripe_Sepa::class;
+					$methods[] = WC_Gateway_Stripe_Bancontact::class;
+					$methods[] = WC_Gateway_Stripe_Sofort::class;
+					$methods[] = WC_Gateway_Stripe_Giropay::class;
+					$methods[] = WC_Gateway_Stripe_Eps::class;
+					$methods[] = WC_Gateway_Stripe_Ideal::class;
+					$methods[] = WC_Gateway_Stripe_P24::class;
+					$methods[] = WC_Gateway_Stripe_Alipay::class;
+					$methods[] = WC_Gateway_Stripe_Multibanco::class;
+					return $methods;
+				}
+
+				if ( $this->is_upe_enabled() ) {
+					$methods[] = WC_Stripe_UPE_Payment_Gateway::class;
 				} else {
-					$methods[] = 'WC_Gateway_Stripe';
-					$methods[] = 'WC_Gateway_Stripe_Sepa';
+					// These payment gateways will be hidden when UPE is enabled:
+					$methods[] = ( class_exists( 'WC_Subscriptions_Order' ) && function_exists( 'wcs_create_renewal_order' ) ) ? WC_Stripe_Subs_Compat::class : WC_Gateway_Stripe::class;
+					$methods[] = WC_Gateway_Stripe_Giropay::class;
+					$methods[] = WC_Gateway_Stripe_Ideal::class;
+					$methods[] = WC_Gateway_Stripe_Bancontact::class;
+					$methods[] = WC_Gateway_Stripe_Eps::class;
 				}
```

Comment:
Isn't this whole `else` statement a duplicate of what's in the `if ( ! WC_Stripe_Features::is_upe_enabled() )` block above?

What's the difference between `WC_Stripe_Features::is_upe_enabled()` and `$this->is_upe_enabled()`?

Reply:
`WC_Stripe_Features::is_upe_enabled()` is if the feature flag is enabled or not. If the feature flag is not enabled, the user will not even see the setting to turn UPE on or off and all of the gateways will load exactly as they were before.

`$this->is_upe_enabled()` on the other hand is whether the merchant has turned UPE on in the Stripe settings. When The feature flag is enabled _and_ the merchant has toggled UPE on in settings, then a different set of gateways are loaded (UPE + any LPM's that aren't supported by UPE).

Reply:
Ah, that makes the `WC_Stripe_Features` class make a bit more sense 😅 
I didn't understand the difference until it was pointed out. Maybe this would be clearer if `WC_Stripe_Features` was instead called `WC_Stripe_Feature_Flags`?

Or if the `WC_Stripe_Features::is_upe_enabled()` was renamed `WC_Stripe_Features::is_upe_flag_enabled()`?

I _was_ worried these were duplicates, but now I'm mostly worried about the readability 🤔 
I think renaming either the class or function would clear up some of the confusion I experienced here.

Reply:
`WC_Stripe_Features` has been renamed to `WC_Stripe_Feature_Flags`.